calculate AbsPath() relative to the configured server base path

jdef opened the pull request:
when apiserver is running behind a proxy w/ a URL prefix (e.g. /services/kubernetes/api) important `kubectl` commands break. this is because `AbsPath()` is completely ignorant of the server URL base. this PR attempts to correct that.

UPDATE: the latest version of this PR...
- requires that the `KUBERNETES_MASTER` envvar (and/or `--server`) **not** end with `/api`
- changes the usage of Config.Host -- if it contains a URL, then any Path segment is considered a proxy-prefix

jdef commented on pkg/client/unversioned/helper.go:
```diff
@@ -208,15 +208,12 @@ func ServerAPIVersions(c *Config) (groupVersions []string, err error) {
 	}
 	// Get the groupVersions exposed at /api
-	baseURL.Path = "/api"
+	baseURL.Path = absPath(prefix, prefix, "/api")
```
oops, missed this

jdef commented on pkg/client/unversioned/helper.go:
```diff
@@ -230,7 +227,7 @@ func ServerAPIVersions(c *Config) (groupVersions []string, err error) {
 	groupVersions = append(groupVersions, v.Versions...)
 	// Get the groupVersions exposed at /apis
-	baseURL.Path = "/apis"
+	baseURL.Path = absPath(prefix, prefix, "/apis")
```
this too

jdef commented:
@jlowdermilk PTAL
/cc @smarterclayton

jdef commented:
hm.. misread blunderbuss. re-assigning to @deads2k

jdef commented:
how can I see the log for the failed unit/integration tests?

smarterclayton commented:
@lavalamp this is the case I mentioned where people want to run Kube under different API endpoints :)

smarterclayton commented:
@jdef you need to copy the e2e results URL, strip off everything after the git commit, then navigate down

smarterclayton commented:
Regarding whether this is supported, I think it should be supported, I think it's fairly normal for people to reparent API resources under arbitrary prefixes in real environments, and I have put various bits of code into places in Kube to enable it.  However, @lavalamp and I have argued about it in the past, so he needs to be part of the agreement to support it.

sttts commented:
From the integration test logs:
```
--- FAIL: TestKubectlValidation (0.14s)
	kubectl_test.go:72: case 3: unexpected error: the server could not find the requested resource
	kubectl_test.go:72: case 4: unexpected error: the server could not find the requested resource
```

sttts commented:
Integration tests are fixed with https://github.com/kubernetes/kubernetes/pull/17535

deads2k commented on pkg/client/unversioned/helper.go:
```diff
@@ -208,15 +208,12 @@ func ServerAPIVersions(c *Config) (groupVersions []string, err error) {
 	}
 	client := http.Client{Transport: transport}
 
-	configCopy := *c
```
This was intended to ensure that we didn't change original, so the `configCopy` should have been first and then used throughout the method.  Rather than deleting it, can you update to match intent.

jdef replied:
done

deads2k commented on pkg/client/unversioned/request.go:
```diff
@@ -232,13 +234,27 @@ func (r *Request) AbsPath(segments ...string) *Request {
+}
+
+func absPath(base string, segments ...string) string {
+	// special case: if base ends in /api then the "real" relative server base is ..
```
Why?  This seems like a really strange thing to do.  Why not choose the base you mean?

deads2k commented on pkg/client/unversioned/request.go:
```diff
@@ -232,13 +234,27 @@ func (r *Request) AbsPath(segments ...string) *Request {
+		return p
 	}
-	return r
+	x := make([]string, 1, len(segments)+1)
```
two `path.Join`s or chained `path.Join(base, path.Join(segments...))` would be easier to read.

deads2k commented on pkg/client/unversioned/helper.go:
```diff
@@ -403,12 +400,12 @@ func RESTClientFor(config *Config) (*RESTClient, error) {
 	}
 
-	client := NewRESTClient(baseURL, config.Version, config.Codec, config.QPS, config.Burst)
+	client := NewRESTClient(baseURL, prefix, config.Version, config.Codec, config.QPS, config.Burst)
```
I would have expected that setting the `baseURL` would have been sufficient.  Are you changing only the `/api` or `/apis` endpoint and not things like `/healthz` for some reason?

jdef replied:
It's my understanding that `Config.Prefix` may point to `.../api` or `.../apis` but not `.../healthz`.

`baseURL` is not a good name. It may actually contain the path to the api version, e.g. `/api/v1`. Probably more aptly named `unversionedBase`.

jdef replied:
ugh. lacking coffee. I meant to say that it would be more aptly named `versionedBase`

jdef commented:
let me clean this up a bit

jdef commented on pkg/client/unversioned/restclient.go:
```diff
@@ -67,6 +69,7 @@ func NewRESTClient(baseURL *url.URL, apiVersion string, c runtime.Codec, maxQPS
 	}
 	return &RESTClient{
 		baseURL:    &base,
+		prefix:     prefix,
```
this is probably better named `serverBase`, and we should rename baseURL to `apiBaseURL`

sttts commented:
```
!!! 'gofmt -s' needs to be run on the following files: 
./pkg/client/unversioned/helper.go
```

sttts commented on pkg/client/unversioned/helper.go:
```diff
@@ -408,7 +421,16 @@ func RESTClientFor(config *Config) (*RESTClient, error) {
+	// baseURL is versioned, let's get a base without the version
+	configCopy := *config
+	configCopy.Version = ""
+	unversioned, err := defaultServerUrlFor(&configCopy)
```
This kind of code repeats multiple times. Maybe it's worth to factor that out somehow?

sttts commented on pkg/client/unversioned/helper.go:
```diff
@@ -592,19 +623,25 @@ func DefaultServerURL(host, prefix, version string, defaultTLS bool) (*url.URL,
-		}
-		hostURL.Path = prefix
-	}
+	prefix = findPrefix(hostURL.Path, prefix)
```
The comment does not fit anymore.

sttts commented on pkg/client/unversioned/helper.go:
```diff
@@ -592,19 +623,25 @@ func DefaultServerURL(host, prefix, version string, defaultTLS bool) (*url.URL,
 }
 
+// findPrefix returns the first non-empty string as the URL prefix. if no such string exists
+// in the provided list, returns "/".
```
findPrefix returns the first non-empty string or "/" if all are empty.

sttts commented on pkg/client/unversioned/helper.go:
```diff
@@ -672,3 +709,18 @@ func TimeoutFromListOptions(options api.ListOptions) time.Duration {
 }
+
+func stripAPIFromPrefix(base string) string {
+	// normalize the base
```
The case apis/ does not happen anymore? Was working around that before in absPath().

jdef replied:
`/apis` is in the `apiPrefixes` map

sttts replied:
I meant with the trailing "/". Didn't follow all the control flow. Maybe it's resolved now.

jdef replied:
shouldn't need to worry about trailing / since paths are Join'd and part of being Join's is being Clean'ed - and that strips a trailing /

sttts commented on pkg/client/unversioned/request.go:
```diff
@@ -232,11 +234,16 @@ func (r *Request) AbsPath(segments ...string) *Request {
+		r.path = p
 	} else {
-		r.path = path.Join(segments...)
+		r.path = path.Join(r.serverBase, path.Join(segments...))
```
path.Join(r.serverBase, sergments...) ?

jdef replied:
nope, bad syntax

sttts replied:
right

sttts commented on pkg/client/unversioned/request.go:
```diff
@@ -232,11 +234,16 @@ func (r *Request) AbsPath(segments ...string) *Request {
 		return r
 	}
+	// join serverBase+segments
 	if len(segments) == 1 {
```
```
r.path = path.Join(r.serverBase, sergments...)
if len(segments) == 1 && strings.HasSuffix(segments[0], "/") && !strings.HasSuffix(r.path, "/") {
  r.path += "/"
}

sttts replied:
maybe simpler

sttts commented:
@jdef the PR name does not fit anymore.

sttts commented:
Well, it's still AbsPath which is changed (next to other things). Maybe ignore my previous comment.

jdef commented:
@deads2k PTAL - have your concerns been addressed?

deads2k commented on pkg/client/unversioned/helper.go:
```diff
@@ -408,7 +419,14 @@ func RESTClientFor(config *Config) (*RESTClient, error) {
 	}
 
-	client := NewRESTClient(baseURL, config.Version, config.Codec, config.QPS, config.Burst)
+	// baseURL is versioned, let's get a base without the version
```
It seems like we should fix our internal structure to be more reasonable and keep the unreasonable `baseURL` contained to the edges where we accept input.

deads2k commented:
I think the internal structs should have attributes that reflect the most reasonable structure.  The reswizzling you've done highlights how difficult that internal structure is to work with.  I would expect the conversion to be contained to just a couple spots.  Can you investigate how reasonable that is?  It would make this change more palatable by making it easier to reason about side-effects or missed locations.

Also, @lavalamp should get a chance to continue his position reference here: https://github.com/kubernetes/kubernetes/pull/17535#issuecomment-158236634

jdef commented:
I'm happy to spend additional time cleaning up unreasonable things. Want to get buy-in from @lavalamp first.

deads2k commented:
@lavalamp I'll take this back, but I want to be sure you know you're the next actor.

jdef commented:
rebased and minor tweaks

jdef commented:
rebased

lavalamp commented:
@smarterclayton I guess if people want to do this, we should make it work...

lavalamp commented on pkg/client/unversioned/helper.go:
```diff
@@ -577,3 +607,26 @@ func TimeoutFromListOptions(options unversioned.ListOptions) time.Duration {
 	return 0
 }
+
+func stripAPIFromPrefix(base string) string {
```
Considering the pain I went through to make the proxy command in kubectl work, I think any solution for which this function is required is going to be a real pain going forward. Instead I suggest:

* Deprecate / start ignoring the "prefix" field in our config file. That was just a mistake to have at all, given the current implementation.
* Add a new field. I'm not sure what to call it (parent-prefix? proxy-prefix?). But it should be blank by default, and if set, it should get prepended to everything, even AbsPath calls.

I think if you do this, this PR will be a lot smaller and simpler.

jdef replied:
Thanks for the feedback, I'll work something up

Commit:
AbsPath should be compatible with proxy-prefixes:
- replace Config.Prefix with .Host and .APIPath
- Request .path promoted to .pathPrefix, .baseURL holds required prefix

jdef commented:
I've removed `Prefix` (for now) and replaced with `ProxyPrefix` and `APIPath`. @lavalamp is this approach closer to what you have in mind? one consequence of this approach is that `APIPath` isn't derived from a config's `Server` variable -- only `ProxyPrefix` is. Is that going to be too limiting in some cases?

jdef commented on pkg/client/unversioned/helper.go:
```diff
@@ -468,12 +483,9 @@ func DefaultServerURL(host, prefix string, groupVersion unversioned.GroupVersion
 
 	// If the user specified a URL without a path component (http://server.com), automatically
-	// append the default prefix
+	// append the proxy prefix
```
... and API-path

jdef commented:
@k8s-bot unit test this

lavalamp commented on cmd/libs/go2idl/client-gen/testoutput/testgroup/unversioned/testgroup_client.go:
```diff
@@ -69,7 +69,7 @@ func setConfigDefaults(config *unversioned.Config) error {
 		return err
 	}
-	config.Prefix = "/apis"
+	config.APIPath = "/apis"
```
I think you need to fix the generator, not modify its output.

jdef replied:
i'm not familiar with the generator - have any pointers?

On Thu, Jan 7, 2016 at 10:11 AM, Daniel Smith <notifications@github.com>
wrote:

> In
> cmd/libs/go2idl/client-gen/testoutput/testgroup/unversioned/testgroup_client.go
> <https://github.com/kubernetes/kubernetes/pull/17535#discussion_r49082111>
> :
>
> > @@ -69,7 +69,7 @@ func setConfigDefaults(config *unversioned.Config) error {
> >  	if err != nil {
> >  		return err
> >  	}
> > -	config.Prefix = "/apis"
> > +	config.APIPath = "/apis"
>
> I think you need to fix the generator, not modify its output.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17535/files#r49082111>.
>

jdef replied:
nevermind - found it

lavalamp commented on pkg/client/unversioned/fake/fake.go:
```diff
@@ -50,23 +50,27 @@ type RESTClient struct {
+	return c.do("DELETE")
+}
+
+func (c *RESTClient) do(verb string) *unversioned.Request {
```
s/do/verb/?

"do" is too similar to the existing Do for my taste

jdef replied:
ack

lavalamp commented on pkg/client/unversioned/helper.go:
```diff
@@ -40,14 +40,29 @@ import (
 )
 
+const (
+	defaultAPIPath        = "/api"
```
legacyAPIPath

lavalamp commented on pkg/client/unversioned/helper.go:
```diff
@@ -40,14 +40,29 @@ import (
 
+const (
+	defaultAPIPath        = "/api"
+	defaultExtensionsPath = "/apis"
```
defaultAPIPath

lavalamp commented on pkg/client/unversioned/helper.go:
```diff
@@ -40,14 +40,29 @@ import (
+		Prefix string
+	*/
+	// ProxyPrefix is a super-path that's prepended to all apiserver requests; optional
+	ProxyPrefix string
```
Can this not just be part of Host?

lavalamp commented on pkg/client/unversioned/helper.go:
```diff
@@ -444,7 +459,7 @@ func UnversionedRESTClientFor(config *Config) (*RESTClient, error) {
 // to use with a Client at a given API version following the standard conventions for a
 // Kubernetes API.
-func DefaultServerURL(host, prefix string, groupVersion unversioned.GroupVersion, defaultTLS bool) (*url.URL, error) {
+func DefaultServerURL(host, proxyPrefix, apiPath string, groupVersion unversioned.GroupVersion, defaultTLS bool) (*url.URL, error) {
```
Can proxyPrefix just be part of host?

jdef replied:
ack

lavalamp commented on pkg/client/unversioned/helper.go:
```diff
@@ -468,12 +483,9 @@ func DefaultServerURL(host, prefix string, groupVersion unversioned.GroupVersion
-			prefix = "/"
-		}
-		hostURL.Path = prefix
+		hostURL.Path = path.Join(proxyPrefix, apiPath)
```
It's not clear to me why apiPath is also appended.

jdef replied:
right, so the plan is to NOT append it here. instead a "versionedAPIPath" will be an additional result returned by this func. hopefully the next refactoring I push makes more sense

lavalamp commented on pkg/client/unversioned/request.go:
```diff
@@ -115,7 +116,7 @@ type Request struct {
 
 // NewRequest creates a new request helper object for accessing runtime.Objects on a server.
-func NewRequest(client HTTPClient, verb string, baseURL *url.URL, groupVersion unversioned.GroupVersion, codec runtime.Codec, backoff BackoffManager) *Request {
+func NewRequest(client HTTPClient, verb string, baseURL *url.URL, proxyPrefix string, groupVersion unversioned.GroupVersion, codec runtime.Codec, backoff BackoffManager) *Request {
```
Here also, can proxyPrefix just be part of baseURL?

lavalamp replied:
(and require that /api or /apis NOT be part of baseURL)

lavalamp commented on pkg/client/unversioned/request.go:
```diff
@@ -244,11 +246,10 @@ func (r *Request) AbsPath(segments ...string) *Request {
 		return r
 	}
-	if len(segments) == 1 {
+	r.path = path.Join(r.proxyPrefix, path.Join(segments...))
```
So with my preferred storage location, this would become `r.path = path.Join(r.baseURL.Path, segments...)`. Will anything go wrong with that?

jdef replied:
well, unit tests are passing when I do that (after refactoring) so that looks promising. going to smoke test a bit and then push my latest changes

karlkfi commented on pkg/client/unversioned/clientcmd/client_config.go:
```diff
@@ -94,7 +94,8 @@ func (config DirectClientConfig) ClientConfig() (*client.Config, error) {
 	if u, err := url.ParseRequestURI(clientConfig.Host); err == nil && u.Opaque == "" && len(u.Path) > 1 {
-		clientConfig.Prefix = u.Path
+		clientConfig.ProxyPrefix = u.Path
+		//TODO(jdef) attempt backwards compat here by checking for /api or /apis and deriving APIPath?
```
is this todo going to be resolved?

jdef replied:
going to kill ProxyPrefix

karlkfi commented on pkg/client/unversioned/helper.go:
```diff
@@ -40,14 +40,29 @@ import (
+		// Prefix is the sub path of the server. If not specified, the client will set
+		// a default value.  Use "/" to indicate the server root should be used
+		Prefix string
+	*/
```
does this comment block need to be nuked?

jdef replied:
yes it will be. this PR isn't final form yet

lavalamp commented:
@caesarxuchao or @krousey can one of you make a pass over this, too?

jdef commented:
@lavalamp PTAL

jdef commented:
@k8s-bot unit test this

krousey commented on pkg/client/unversioned/fake/fake.go:
```diff
@@ -50,23 +50,27 @@ type RESTClient struct {
+}
+
+func (c *RESTClient) request(verb string) *unversioned.Request {
+	return unversioned.NewRequest(c, verb, &url.URL{Host: "localhost"}, "", *testapi.Default.GroupVersion(), c.Codec, nil)
```
This is a nice cleanup, but please try to keep PRs focused and singular in purpose in the future.

jdef replied:
message received. only cleaned it up because i was really annoyed with the number of changes I had to make because of the prior copy/paste

krousey replied:
Oh yea, this did change. See how I didn't notice that. :smile_cat:

lavalamp replied:
(fwiw I approve of the change-- it puts it back the way I originally had it ;) )

krousey commented on pkg/client/unversioned/helper.go:
```diff
@@ -460,32 +465,25 @@ func DefaultServerURL(host, prefix string, groupVersion unversioned.GroupVersion
-		}
-		hostURL.Path = prefix
-	}
+	versionedAPIPath := path.Join("/", hostURL.Path, apiPath)
```
This is confusing to me. I would expect `hostURL.Path` to be empty since it should just be the host. What does it mean if `host` is "example.com/prefix_one/" and `apiPath` is "prefix_two"?

jdef replied:
hostURL.Path contains an optional proxy-prefix. apiPath is generally either
/api or /apis -- it's a pointer to the API entrypoint of the server (where
/api is legacy and /apis is the future). i'd love to get rid of apiPath but
I just don't see how to do that while we need to support both. suggestions?

On Thu, Jan 7, 2016 at 5:36 PM, krousey <notifications@github.com> wrote:

> In pkg/client/unversioned/helper.go
> <https://github.com/kubernetes/kubernetes/pull/17535#discussion_r49136318>
> :
>
> >  		}
> >  	}
> >
> > -	// If the user specified a URL without a path component (http://server.com), automatically
> > -	// append the default prefix
> > -	if hostURL.Path == "" {
> > -		if prefix == "" {
> > -			prefix = "/"
> > -		}
> > -		hostURL.Path = prefix
> > -	}
> > +	versionedAPIPath := path.Join("/", hostURL.Path, apiPath)
>
> This is confusing to me. I would expect hostURL.Path to be empty since it
> should just be the host. What does it mean if host is "
> example.com/prefix_one/" and apiPath is "prefix_two"?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17535/files#r49136318>.
>

lavalamp replied:
Please add a comment saying what hostURL.Path contains.

jdef replied:
ack

lavalamp commented on cmd/libs/go2idl/client-gen/generators/generator-for-group.go:
```diff
@@ -157,7 +157,7 @@ func setConfigDefaults(config *$.Config|raw$) error {
 	if err != nil {
 		return err
 	}
-	config.Prefix = $.prefix$
```
Can you hunt down where prefix is set/used and change the name to apiPath?

jdef replied:
done

lavalamp commented on pkg/client/unversioned/helper.go:
```diff
@@ -40,14 +40,18 @@ import (
-	// a default value.  Use "/" to indicate the server root should be used
-	Prefix string
+	// APIPath is a sub-path that points to an API root.
+	APIPath string
```
@caesarxuchao is this struct ever serialized to disk?

lavalamp commented on pkg/client/unversioned/request.go:
```diff
@@ -244,11 +244,10 @@ func (r *Request) AbsPath(segments ...string) *Request {
 		return r
 	}
-	if len(segments) == 1 {
+	r.path = path.Join(r.baseURL.Path, path.Join(segments...))
```
This is inconsistent with line 128-- there, r.path only has the api part. Here, it will also include the proxy prefix. Won't this result in doubling the proxy prefix?

jdef replied:
on line 127: r.baseURL.Path includes whatever proxy-prefix was specified. no /api, /apis, or version info. this is a semantic change that I've introduced.

on line 128: r.path defaults to versionedAPIPath, which includes the proxy prefix as well as (optionally) a path to an API version (e.g. /api/v1). it's a "fully-qualified path" that's used to access API resources. i haven't changed the semantics of this variable, it's always been the full path to an API variant.

here (247): we're saying "build a path that begins with the proxy-prefix (if there is one) append segments to it"; there's no duplication. This is change from the prior behavior that would wipe out any desired proxy-prefix.

lavalamp replied:
That's not obvious-- I added a comment there.

lavalamp commented on pkg/client/unversioned/helper.go:
```diff
@@ -40,14 +40,18 @@ import (
 // Config holds the common attributes that can be passed to a Kubernetes client on
 // initialization.
 type Config struct {
 	// Host must be a host string, a host:port pair, or a URL to the base of the API.
```
Please extend this comment.

jdef replied:
ack

jdef commented:
@lavalamp I've address your feedback

lavalamp commented on pkg/client/unversioned/request.go:
```diff
@@ -115,7 +115,7 @@ type Request struct {
 
 // NewRequest creates a new request helper object for accessing runtime.Objects on a server.
-func NewRequest(client HTTPClient, verb string, baseURL *url.URL, groupVersion unversioned.GroupVersion, codec runtime.Codec, backoff BackoffManager) *Request {
+func NewRequest(client HTTPClient, verb string, baseURL *url.URL, versionedAPIPath string, groupVersion unversioned.GroupVersion, codec runtime.Codec, backoff BackoffManager) *Request {
```
If versionedAPIPath is supposed to overlap with any prefix in baseURL, please a) don't do that, or b) document document document, and I prefer a) unless there's a super good reason?

jdef replied:
OK, I've made versionedAPIPath relative

jdef commented:
@lavalamp I've addressed your concerns re: making versionedAPIPath relative. PTAL

lavalamp commented on pkg/client/unversioned/request.go:
```diff
@@ -115,17 +115,22 @@ type Request struct {
 		verb:         verb,
 		baseURL:      baseURL,
-		path:         baseURL.Path,
+		path:         path.Join(pathPrefix, versionedAPIPath),
```
I guess I don't get why you store path with the prefix. Why not just join baseURL and path at the very end when you're about to make the request, and keep them separate until then?

(sorry for slow responses; I have limited availability last week and this week)

jdef replied:
`path` is really a bad name for this variable. it's actually treated as a "path prefix" (see https://github.com/mesosphere/kubernetes/blob/jdef_abspath_breaks_proxies/pkg/client/unversioned/request.go#L143)

so I think we're actually doing the right thing here. i'm happy to rename this for clarity -- would that satisfy?

lavalamp replied:
OK, I see. Yes, renaming it would make me happy. Sorry for delay, I'm back in the office now.

jdef replied:
done

jdef commented:
squashed

jdef commented:
@lavalamp PTAL - I've addressed all review feedback and squashed

lavalamp commented:
I will try to make a final pass tomorrow. Sorry for the delay.

jdef commented:
Sounds good - thanks!

On Tue, Jan 19, 2016 at 8:03 PM, Daniel Smith <notifications@github.com>
wrote:

> I will try to make a final pass tomorrow. Sorry for the delay.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17535#issuecomment-173039940>
> .
>

lavalamp commented:
LGTM thanks for bearing with me

jdef commented:
@lavalamp :+1: no worries, thanks for jumping on this your first day back
